src/main/java/uk/ac/ebi/eva/pipeline/io/writers/AnnotationMongoWriter.java

```diff
@@ -0,0 +1,143 @@
+
+    private static final String ANNOTATION_XREF_ID_FIELD = Annotation.XREFS_FIELD + "." + Xref.XREF_ID_FIELD;
+
+    private static final String ANNOTATION_CT_SO_FIELD = Annotation.CONSEQUENCE_TYPE_FIELD + "."
+            + ConsequenceType.SO_ACCESSION_FIELD;
+    public static final String EACH = "$each";
+    public static final String ADD_TO_SET = "$addToSet";
+
+    private final MongoOperations mongoOperations;
+
+    private final String collection;
+
+    public AnnotationMongoWriter(MongoOperations mongoOperations, String collection) {
+        Assert.notNull(mongoOperations);
+        Assert.hasText(collection);
+        this.mongoOperations = mongoOperations;
+        this.collection = collection;
+
+        createIndexes();
+    }
+
+    @Override
+    public void write(List<? extends Annotation> annotations) throws Exception {
+        BulkOperations bulk = mongoOperations.bulkOps(BulkOperations.BulkMode.UNORDERED, collection);
+        prepareBulk(annotations, bulk);
+        bulk.execute();
+    }
+
+    private void prepareBulk(List<? extends Annotation> annotations, BulkOperations bulk) {
+        Map<String, Annotation> annotationsByStorageId = groupAnnotationById(annotations);
+        for (Annotation annotation : annotationsByStorageId.values()) {
+            writeAnnotationInMongoDb(bulk, annotation);
+        }
+    }
+
+    private Map<String, Annotation> groupAnnotationById(List<? extends Annotation> annotations) {
+        Map<String, Annotation> groupedAnnotations = new HashMap<>();
+        for (Annotation annotation : annotations) {
+            String id = annotation.getId();
+            groupedAnnotations.computeIfPresent(id, (key, oldVar) -> oldVar.concatenate(annotation));
+            groupedAnnotations.putIfAbsent(id, annotation);
+        }
+        return groupedAnnotations;
+    }
+
+    private void writeAnnotationInMongoDb(BulkOperations bulk, Annotation annotation) {
+        Query upsertQuery = new BasicQuery((DBObject)convertToMongo(new SimplifiedAnnotation(annotation)));
+        Update update = buildUpdateQuery(annotation);
+        bulk.upsert(upsertQuery, update);
+    }
+
+    private BasicUpdate buildUpdateQuery(Annotation annotation) {
+        final BasicDBObject addToSetValue = new BasicDBObject();
+        addToSetValue.append(CONSEQUENCE_TYPE_FIELD, buildInsertConsequenceTypeQuery(annotation));
+        addToSetValue.append(XREFS_FIELD, buildInsertXrefsQuery(annotation));
+        return new BasicUpdate(new BasicDBObject(ADD_TO_SET, addToSetValue));
+    }
+
+    private BasicDBObject buildInsertXrefsQuery(Annotation annotation) {
+        return new BasicDBObject(EACH, convertToMongo(annotation.getXrefs()));
+    }
+
+    private BasicDBObject buildInsertConsequenceTypeQuery(Annotation annotation) {
+        return new BasicDBObject(EACH, convertToMongo(annotation.getConsequenceTypes()));
+    }
+
+    private Object convertToMongo(Object object) {
```

Comment:
Agreed offline to simplify this call by forcing the method to receive a `SimplifiedAnnotation` and return an already casted DBObject.